pkg/scheduler/priorities.go

```diff
@@ -132,3 +134,92 @@ func (n *NodeLabelPrioritizer) CalculateNodeLabelPriority(pod api.Pod, podLister
 	}
 	return result, nil
 }
+
+// BalancedResourceAllocation is favors nodes with balanced resource usage rate.
+// BalancedResourceAllocation should **NOT** be used alone, and **MUST** be used together with LeastRequestedPriority.
+// It calculates the percentage of memory and CPU requested over capacity, and prioritizes based on how close the
+// two percentage metrics are to each other.
+// Details: score = 1 / sqrt[(cpu_pct - average)^2 + (mem_pct - average)^2], average = (cpu_pct + mem_pct) / 2
+func BalancedResourceAllocation(pod api.Pod, podLister PodLister, minionLister MinionLister) (HostPriorityList, error) {
+	nodes, err := minionLister.List()
+	if err != nil {
+		return HostPriorityList{}, err
+	}
+	podsToMachines, err := MapPodsToMachines(podLister)
+
+	list := HostPriorityList{}
+	for _, node := range nodes.Items {
+		list = append(list, calculateBalancedResourceAllocation(pod, node, podsToMachines[node.Name]))
+	}
+	return list, nil
+}
+
+func calculateBalancedResourceAllocation(pod api.Pod, node api.Node, pods []api.Pod) HostPriority {
+	totalMilliCPU := int64(0)
+	totalMemory := int64(0)
+	score := int(0)
+	for _, existingPod := range pods {
+		for _, container := range existingPod.Spec.Containers {
+			totalMilliCPU += container.Resources.Limits.Cpu().MilliValue()
+			totalMemory += container.Resources.Limits.Memory().Value()
+		}
+	}
+	// Add the resources requested by the current pod being scheduled.
+	// This also helps differentiate between differently sized, but empty, minions.
+	for _, container := range pod.Spec.Containers {
+		totalMilliCPU += container.Resources.Limits.Cpu().MilliValue()
+		totalMemory += container.Resources.Limits.Memory().Value()
+	}
+
+	capacityMilliCPU := node.Status.Capacity.Cpu().MilliValue()
+	capacityMemory := node.Status.Capacity.Memory().Value()
+
+	cpuPctOfCapRequested := calculatePctOfCapRequested(totalMilliCPU, capacityMilliCPU, node.Name)
+	memoryPctOfCapRequested := calculatePctOfCapRequested(totalMemory, capacityMemory, node.Name)
+	avergePctOfCapRequested := (cpuPctOfCapRequested + memoryPctOfCapRequested) / 2
+	cpuDiff := cpuPctOfCapRequested - avergePctOfCapRequested
+	memoryDiff := memoryPctOfCapRequested - avergePctOfCapRequested
+	diff := math.Sqrt(math.Pow(cpuDiff, 2) + math.Pow(memoryDiff, 2))
+	// closer percentage metrics lead to higher score with 10 being the highest
```

Comment:
Hi @lavalamp

I appreciate your comments here very much and I am going to change them accordingly. About the proposal of the scaling mechanism, here are some of my thoughts:

1. I agree with you that there are many more gentle ways to deal with difference other than that in the current patch. 
2. As long as the score is capable to prefer similar fraction of capability, I don't think we could miss anything essential here. 
3. Given that the LeastRequestedPriority function is also a "gentle" method, I think we'd better off doing the same here.

However, I am not sure that I have fully understood your proposal, and to the best of my understanding(should be very similar to yours), here is what I am going to change it:

1. consider the difference between cpuFraction and memoryFraction directly, the lower and upper bound are (-1.0) ~ (1.0)
2. use absolute value here and scale to 0-10: |cpuFraction-memoryFraction| * 10
3. subtract from 10 to have the score whose graph is like this: https://www.wolframalpha.com/input/?i=graph+10-%7Cx-y%7C*10%2C+x+in+%280%2C1%29%2C+y+in+%280%2C+1%29, which is gentle and intuitive enough to show the essence of this priority function.

note: the computation could start from calculating the averageFraction as well, but it would end up with the same solution as above.

I am going to change scale mechanism this way. What do you think?

PS: I will squash them as well.